File: Trie.py


class TrieNode:
    children = {}
    isTerminating = False
    def __init__ (self, value = None, parent = None):
        self.value = value
        self.parent = parent
    def addChild(self, child):
        if child in self.children:
            return
        self.children[child] = TrieNode(child, None)


class Trie:
    def __init__(self):
        self._root = TrieNode()

    def Insert(self, word):
        if not word:
            return

        currentNode = self._root
        currentIndex = 0

        while currentIndex < len(word):
            char = word[currentIndex]
            if char in currentNode.children:
                currentNode = currentNode.children[char]
            else:
                currentNode.addChild(char)
                currentNode = currentNode.children[char]

            currentIndex += 1
        currentNode.isTerminating = True

    def Contains(self, word):
        if not word:
            return

        currentNode = self._root
        currentIndex = 0

        while currentIndex < len(word):
            char = word[currentIndex]
            if char in currentNode.children:
                currentNode = currentNode.children[char]
                currentIndex += 1
                if currentIndex == len(word) and currentNode.isTerminating:
                    return True

        return False

File: Trie.swift
import Foundation

class TrieNode<T: Hashable> {
    var value: T?
    weak var parent: TrieNode?
    var isTerminating = false
    var children : [T: TrieNode] = [:]
    
    init(value: T? = nil, parent: TrieNode? = nil) {
        self.value = value
        self.parent = parent
    }
    
    func add(child: T) {
        guard children[child] == nil else { return }
        
        children[child] = TrieNode(value: child, parent: nil)
    }
}

class Trie {
    typealias Node = TrieNode<Character>
    fileprivate let root : Node
    
    init() {
        root = Node()
    }
    
    func Insert(word: String) {
        guard !word.isEmpty else { return }
        
        var currentNode = root
        
        let characters = Array(word.lowercased().characters)
        var currentIndex = 0
        
        while currentIndex < characters.count {
            let character = characters[currentIndex]
            
            if let child = currentNode.children[character] {
                currentNode = child
            } else {
                currentNode.add(child: character)
                currentNode = currentNode.children[character]!
            }
            currentIndex += 1
        }
        currentNode.isTerminating = true
    }
    
    func Contains(word: String) -> Bool {
        guard !word.isEmpty else { return false}
        
        var currentNode = root
        let characters = Array(word.lowercased().characters)
        var currentIndex = 0
        while currentIndex < characters.count , let child = currentNode.children[characters[currentIndex]] {
            
            currentIndex += 1
            currentNode = child
        }
        if currentIndex == characters.count && currentNode.isTerminating {
            return true
        }
        return false
    }
    
    func findMax() -> String {
        
    }
}


let trie = Trie()
let file = "wordsforproblem.txt" //this is the file. we will write to and read from it


if let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
    
    let path = dir.appendingPathComponent(file)
    print(dir)
    //reading
    do {
        let data = try String(contentsOf: path, encoding: String.Encoding.utf8)
        let content = data.components(separatedBy: .newlines)
        for word in content {
            trie.Insert(word: word)
        }
    }
    catch {/* error handling here */}
}

print(trie.findLogest())

